### ISSUE TITLE
OSQuery node (tls plugin) stops executing if it can't enroll for a while

### ISSUE THREAD

**Issue by casimanu** [osquery/osquery#4433]
### Description
When I run an OSQuery node (using tls plugin) and the server responds, in the enrollment, with "invalid_node": True several times, the node shows an error related to the logger plugin and stop execution.
It's important to mention that if the OSQuery node was running and then, the server starts failing in the enrollment sending "invalid_node": True, the problem doesn't show up. This problem happens only when the node is started and the server fails in the enrollment.

### What version of `osquery` are you using? 

I tested this over two different environments.
```
osquery_version = 2.11.0
os_version = 16.04.4 LTS (Xenial Xerus)
os_build = 
os_platform = ubuntu
```
and

```
osquery_version = 3.2.4
os_version = 10.0.14393
os_build = 14393
os_platform = windows
```

### What have you already tried?

I created a zip with the files I used for reproducing the problem: [2. osquery node stop execution after failing several times enrollment.zip](https://github.com/facebook/osquery/files/2040563/2.osquery.node.stop.execution.after.failing.several.times.enrollment.zip)
I also added a demo video.

I found this in this way: I had a server running and, at some point, it started failing in the enrollment. I decided to re-start the OSQuery node and after failing several attempts to enroll the node showed an error and stopped execution.

Steps to reproduce:

1. Go to server folder
2. Run in a terminal/console a very slightly modified version of the Basic Test Python Server doing: 

```
python test_http_server_not_enroll.py --tls --persist --verbose --cert "../certs/ors.pem" --key "../certs/ors.key" --ca "../certs/myCA.pem" --enroll_secret "../shared-key.txt" 8003
```
3. Type `test` for the PEM pass phrase

4. Go to osquery-node folder
5. run ./osquery.sh (or osquery.bat in Windows) in a other terminal/console
6. Wait for near a minute

### What did you expect to see?

The OSQuery node should continue trying to reach the server for enrolling in the same way if it wasn't restarted it.

### What did you see instead?

After near a minute the OSQuery node execution is stopped with the follow error:
```
"E0524 15:21:18.626893 22424 init.cpp:545] Cannot activate tls logger plugin: No node key, TLS logging disabled."
```
For more details see `error.log` file in the zip file attached.

### Additional Comments:
 - If the OSQuery node doesn't include the logger/config plugins this problem doesn't happen.

**Comment by casimanu** [comment]
I tested these steps to reproduce with the new OSQuery 3.2.6 over Windows and the problem is still happening.

**Comment by Rich5** [comment]
Just wanted to bump this issue. We're experience behavior where if a Windows agent can't enroll for some time it seems to just stop trying.

**Comment by groob** [comment]
osquery uses a flag to control the number of enrollment attempts before it quits.
```
    --config_tls_max_attempts VALUE           Number of attempts to retry a TLS config/enroll request
```

**Comment by casimanu** [comment]
The `--config_tls_max_attempts` value is the max amount of attempts that the same request (config, enroll) is made if they are not successful. After that, some time is waited and a new request (config, enroll) is tried.

If you debug the code you can see this very clear.

The problem I'm mentioned here is that after 1 or 2 minutes of this cycle (in which the `--config_tls_max_attempts` was reached maybe 10 times) the node throws an error related to the logger plugin and it stops working. And this happens only if you set your node to work with scheduled queries (config, logger) because if you set your node to use only ad-hoc queries (distributed_read, distributed_write) the problem doesn't show up. In this last case, the node continues working in this cycle I mentioned forever.

**Comment by mcarpenterjr** [comment]
Do you have an entry like this `SSL SHORT_READ_ERROR: http_client closing socket` in your status log? This is what I'm finding with Windows agents using the TLS Plugin before they go dark. This could relate to #4235

**Comment by casimanu** [comment]
yes @mcarpenterjr.

This are my last lines:
`I0525 15:42:55:691172 13112 http_client.cpp:51] SSL SHORT_READ_ERROR: http_client closing socket
E0524 15:30:13.945073 23098 init.cpp:545] Cannot activate tls logger plugin: No node key, TLS logging disabled.
`

**Comment by mcarpenterjr** [comment]
There are a number of different servers i have in use that auto update, our internal infrastructure is non-scaling so I can only really speculate that an update overnight or during down time would break the communication link in my case. I did find that for osquery running as a service on windows if you open services as an Administrator and open the properties of the service. On the recovery tab set first and second failures to restart. Fleet has remained in contact with the windows host in question for the past 3 days which is an improvement over dropping out after 18 hours or so.

If this continues to become an issue i guess as a last resort you could use the third failure block to run a script that kills off any remaining osquery processes and then restarts the service. You could even do this at the first failure and if you wanted to get trick record the failure in a log and send it to an aggregator if you have one setup.

**Comment by casimanu** [comment]
I checked this issue over:

```
osquery_version = 3.3.2
os_version = 18.10 (Cosmic Cuttlefish)
os_build = 
os_platform = ubuntu
```
and it's still happening.

**Comment by theopolis** [comment]
I am pretty sure this has been fixed in the 4.4.0+ series. Please let us know if it is still an issue.
